library/std/src/sys/windows/process.rs

```diff
@@ -682,6 +706,57 @@ fn make_dirp(d: Option<&OsString>) -> io::Result<(*const u16, Vec<u16>)> {
     }
 }
 
+/// Wrapper around a u8 buffer which ensures that a ProcThreadAttributeList
+/// is not moved during use and is not freed before calling Delete.
+struct ProcThreadAttributeList(Pin<Vec<u8>>);
+impl Drop for ProcThreadAttributeList {
+    fn drop(&mut self) {
+        unsafe { c::DeleteProcThreadAttributeList(self.0.as_mut_ptr().cast()) }
+    }
+}
+/// Wrapper around the value data to be used as a Process Thread Attribute.
+/// This exists primarily to force the raw pointer type to be `Send` and `Sync`
+/// without needing to `unsafe impl` them for `Command`.
+#[derive(Copy, Clone)]
+struct ProcThreadAttributeValue {
+    ptr: *mut c_void,
+    size: usize,
+}
+unsafe impl Send for ProcThreadAttributeValue {}
+unsafe impl Sync for ProcThreadAttributeValue {}
+
+fn make_proc_thread_attributes(
+    attributes: &BTreeMap<usize, ProcThreadAttributeValue>,
+) -> io::Result<ProcThreadAttributeList> {
+    let mut buffer_size = 0;
+    let count = attributes.len() as u32;
+    // First call to get required size. Error is expected.
+    unsafe { c::InitializeProcThreadAttributeList(ptr::null_mut(), count, 0, &mut buffer_size) };
+    let mut buf = Pin::new(crate::vec::from_elem(0u8, buffer_size as usize));
```

Comment:
Does this need to be `Pin`? Surely the `Vec` itself can be moved around freely?

Reply:
The `Pin` prevents us from calling (for example) vec.push() or doing anything that may re-allocate (and move) the underlying data. I'm not 100% sure it is necessary to not move the data, as I don't know the internal workings of these Initialize/Update/Delete functions, but it seemed the right choice to me since there is at least no good reason to move or grow the data after that point.

The `Pin` doesn't prevent us from moving the Vec itself (after all, we must move it out of this function to return it), just the data it points to.

Reply:
> Hm, it does feel like `Command` should ideally take ownership of the value somehow.

I think I agree. I am considering making the method generic over a `T: Sized + Any`, in which case we could grab the size ourselves and store the `T` in some `Box<dyn Any>`, making it slightly easier on the user. A couple concerns I have about that are:

- Do we really need to allocate for every attribute?
- Do we really need ownership of the generic type? Are there any situations where the user might need to retain ownership?
    - I'm *guessing* that any value passed in here would always be `Copy` so this shouldn't be an issue, but I'm not certain of that.

That said, the more I think about it, the more I tend to think taking ownership would be the right choice.

> Other than that it would be good to have safe (or at least safer) APIs

I'm skeptical this could ever be made fully safe, however it will certainly be possible (and I'd say, encouraged) to build safe APIs on top of this.

The reason I'm skeptical basically comes down to the variety of different attributes a user could pass in and the fact that MS could add new attributes at any time. For instance, if MS added something which allowed the user to have some of the parent process's memory mapped into the child's address space, that would immediately preclude this from ever being truly safe.

My hope is that this extension method, it will enable authors of 3rd party libraries to make safe APIs via their own extension traits, even if those safe APIs never make it into the standard library.

Reply:
`u8` is `Unpin`, so `Pin<Vec<u8>>` so both `Pin::into_inner` and `DerefMut` are implemented allowing you to move the inner bytes anyway.

https://doc.rust-lang.org/stable/std/pin/struct.Pin.html#method.into_inner

Reply:
I think perhaps a boxed slice would be enough to prevent accidentally reallocating? Or it might be a good idea to fully contain the owned data. So have an API like:

```rust
struct ProcThreadAttributeList(Vec<u8>); // or Box<[T]>
impl ProcThreadAttributeList {
    fn new(attribute_count: u32) -> io::Result<Self> { ... }
    unsafe fn update(&mut self, attribute: usize, value_ptr: *mut c_void, value_size: u32) -> io::Result<()> { ... }
}
impl Drop for ProcThreadAttributeList { ... }
```

That way there is no reason to touch the inner data.

Reply:
@ChrisDenton I could do that, just seemed like overkill to me when I was writing it.

@bjorn3 You could copy the bytes, or overwrite them, but you can't move them. `let mut v = Pin::new(vec![]); v.push(());` does not compile. Unless I'm just being really dumb right now. You are right about `Pin::into_inner` but calling that explicitly indicates it is now okay to move the underlying data (which it might be anyway, but, again, I'm not certain).